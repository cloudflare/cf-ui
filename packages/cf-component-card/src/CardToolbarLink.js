import React from 'react';
import PropTypes from 'prop-types';
import Link from 'cf-component-link';
import { createComponent } from 'cf-style-container';

const styles = ({ isActive }) => ({
  display: 'inline-block',
  marginRight: '1em',

  '&::after': {
    display: 'inline-block',
    content: "'\\f003'",
    fontFamily: '"cloudflare-font"',
    fontStyle: 'normal',
    fontWeight: 'normal',

    transition: 'transform 0.2s ease',
    transform: isActive ? 'rotate(90deg)' : 'initial'
  }
});

class CardToolbarLink extends React.Component {
  constructor(props) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
  }

  handleClick(e) {
    e.preventDefault();
    this.props.onClick();
  }

  render() {
    return (
      <Link
        role="tab"
        id={this.props.id}
        className={this.props.className}
        onClick={this.handleClick}
      >
        {this.props.children}
      </Link>
    );
  }
}

CardToolbarLink.propTypes = {
  className: PropTypes.string,
  onClick: PropTypes.func.isRequired,
  isActive: PropTypes.bool.isRequired,
  id: PropTypes.string,
  children: PropTypes.node
};

export default createComponent(styles, CardToolbarLink, ['isActive']);
