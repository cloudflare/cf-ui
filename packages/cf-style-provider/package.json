{
  "name": "cf-style-provider",
  "description": "Cloudflare Style Provider",
  "version": "2.0.1",
  "main": "lib/index.js",
  "module": "es/index.js",
  "jsnext:main": "es/index.js",
  "author": "Vojtech Miksu <vojtech@cloudflare.com>",
  "license": "BSD-3-Clause",
  "publishConfig": {
    "registry": "http://registry.npmjs.org/"
  },
  "dependencies": {
    "cf-style-const": "^1.6.2",
    "cf-style-container": "^1.3.2",
    "fela": "^5.0.0",
    "fela-beautifier": "^5.0.0",
    "fela-monolithic": "^5.0.2",
    "fela-plugin-embedded": "^5.0.0",
    "fela-plugin-fallback-value": "^5.0.0",
    "fela-plugin-lvha": "^5.0.0",
    "fela-plugin-named-media-query": "^5.0.1",
    "fela-plugin-prefixer": "^5.0.0",
    "fela-plugin-unit": "^5.0.0",
    "fela-plugin-validator": "^5.0.0",
    "inline-style-prefixer": "^3.0.5",
    "prop-types": "^15.5.8",
    "react-fela": "^5.0.0",
    "react-test-renderer": "^15.6.1"
  },
  "peerDependencies": {
    "react": "^0.14.9 || ^15.0.0-0 || ^16.0.0-0"
  }
}
