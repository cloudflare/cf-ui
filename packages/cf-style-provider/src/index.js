import React, {
  Component,
  PropTypes,
  Children,
  isValidElement,
  cloneElement
} from 'react';
import { createRenderer as createFelaRenderer } from 'fela';
import prefixer from 'fela-plugin-prefixer';
import fallbackValue from 'fela-plugin-fallback-value';
import unit from 'fela-plugin-unit';
import lvha from 'fela-plugin-lvha';
import fontRenderer from 'fela-font-renderer';
import validator from 'fela-plugin-validator';
import beautifier from 'fela-beautifier';
import monolithic from 'fela-monolithic';
import { Provider } from 'react-fela';
import { variables } from 'cf-style-const';
import { ThemeProvider } from 'cf-style-container';

const defaultOpts = {
  selectorPrefix: 'cf-',
  dev: false,
  fontNode: undefined,
  cssNode: undefined
};

export const createRenderer = opts => {
  const usedOpts = Object.assign({}, defaultOpts, opts);
  const plugins = [prefixer(), fallbackValue(), unit(), lvha()];
  const enhancers = [fontRenderer(usedOpts.fontNode)];

  if (usedOpts.dev === true) {
    plugins.push(validator());
    enhancers.push(beautifier());
    enhancers.push(monolithic());
  }

  return createFelaRenderer({
    plugins,
    enhancers,
    selectorPrefix: [usedOpts.selectorPrefix]
  });
};

export const StyleProvider = ({
  selectorPrefix,
  dev,
  cssNode,
  fontNode,
  children,
  ...restProps
}) => {
  const renderer = createRenderer({
    selectorPrefix,
    dev,
    fontNode
  });
  const child = Children.only(children);
  return (
    <Provider renderer={renderer} mountNode={cssNode}>
      <ThemeProvider theme={variables}>
        {isValidElement(child) ? cloneElement(child, { ...restProps }) : child}
      </ThemeProvider>
    </Provider>
  );
};

StyleProvider.defaultProps = defaultOpts;
StyleProvider.propTypes = {
  dev: React.PropTypes.bool,
  selectorPrefix: React.PropTypes.string,
  cssNode: React.PropTypes.object,
  fontNode: React.PropTypes.object,
  children: PropTypes.node.isRequired
};
