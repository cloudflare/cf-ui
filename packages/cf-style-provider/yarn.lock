# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


asap@~2.0.3:
  version "2.0.5"
  resolved "https://registry.yarnpkg.com/asap/-/asap-2.0.5.tgz#522765b50c3510490e52d7dcfe085ef9ba96958f"

bowser@^1.6.0:
  version "1.6.1"
  resolved "https://registry.yarnpkg.com/bowser/-/bowser-1.6.1.tgz#9157e9498f456e937173a2918f3b2161e5353eb3"

core-js@^1.0.0:
  version "1.2.7"
  resolved "https://registry.yarnpkg.com/core-js/-/core-js-1.2.7.tgz#652294c14651db28fa93bd2d5ff2983a4f08c636"

css-in-js-utils@^1.0.3:
  version "1.0.3"
  resolved "https://registry.yarnpkg.com/css-in-js-utils/-/css-in-js-utils-1.0.3.tgz#9ac7e02f763cf85d94017666565ed68a5b5f3215"
  dependencies:
    hyphenate-style-name "^1.0.2"

cssbeautify@0.3.1:
  version "0.3.1"
  resolved "https://registry.yarnpkg.com/cssbeautify/-/cssbeautify-0.3.1.tgz#12dd1f734035c2e6faca67dcbdcef74e42811397"

encoding@^0.1.11:
  version "0.1.12"
  resolved "https://registry.yarnpkg.com/encoding/-/encoding-0.1.12.tgz#538b66f3ee62cd1ab51ec323829d1f9480c74beb"
  dependencies:
    iconv-lite "~0.4.13"

fbjs@^0.8.4:
  version "0.8.12"
  resolved "https://registry.yarnpkg.com/fbjs/-/fbjs-0.8.12.tgz#10b5d92f76d45575fd63a217d4ea02bea2f8ed04"
  dependencies:
    core-js "^1.0.0"
    isomorphic-fetch "^2.1.1"
    loose-envify "^1.0.0"
    object-assign "^4.1.0"
    promise "^7.1.1"
    setimmediate "^1.0.5"
    ua-parser-js "^0.7.9"

fela-beautifier@^4.3.2:
  version "4.3.2"
  resolved "https://registry.yarnpkg.com/fela-beautifier/-/fela-beautifier-4.3.2.tgz#4612902481216d2ac941e18f38166bc0cc76e38c"
  dependencies:
    cssbeautify "0.3.1"

fela-font-renderer@^4.3.2:
  version "4.3.2"
  resolved "https://registry.yarnpkg.com/fela-font-renderer/-/fela-font-renderer-4.3.2.tgz#d0afe027f69209c1afd7403f28787e2c2bd7ac64"

fela-monolithic@^4.3.2:
  version "4.3.2"
  resolved "https://registry.yarnpkg.com/fela-monolithic/-/fela-monolithic-4.3.2.tgz#bf90f4b4fb9326bdc910cf7a0d15cce8a001b697"
  dependencies:
    css-in-js-utils "^1.0.3"

fela-plugin-fallback-value@^4.3.2:
  version "4.3.2"
  resolved "https://registry.yarnpkg.com/fela-plugin-fallback-value/-/fela-plugin-fallback-value-4.3.2.tgz#e1e7cb3ffc992f5898265d6a9f06f8eb325060e7"
  dependencies:
    css-in-js-utils "^1.0.3"

fela-plugin-lvha@^4.3.2:
  version "4.3.2"
  resolved "https://registry.yarnpkg.com/fela-plugin-lvha/-/fela-plugin-lvha-4.3.2.tgz#5d9920f3bdb9f96be11c0744b658e70a14d8774f"

fela-plugin-prefixer@^4.3.2:
  version "4.3.2"
  resolved "https://registry.yarnpkg.com/fela-plugin-prefixer/-/fela-plugin-prefixer-4.3.2.tgz#118715b553d65212ce523935e767a83383dcc257"
  dependencies:
    css-in-js-utils "^1.0.3"
    inline-style-prefixer "^3.0.0"

fela-plugin-unit@^4.3.2:
  version "4.3.2"
  resolved "https://registry.yarnpkg.com/fela-plugin-unit/-/fela-plugin-unit-4.3.2.tgz#278260ad793cf97045f6fdf154bef80c7f0b4445"
  dependencies:
    css-in-js-utils "^1.0.3"

fela-plugin-validator@^4.3.2:
  version "4.3.2"
  resolved "https://registry.yarnpkg.com/fela-plugin-validator/-/fela-plugin-validator-4.3.2.tgz#491f52c0e94cd36c4dccc04013a5ee286e0963ff"

fela@^4.3.2:
  version "4.3.2"
  resolved "https://registry.yarnpkg.com/fela/-/fela-4.3.2.tgz#1e407351bf26d618665bc1bc19708cfbce78bd03"
  dependencies:
    css-in-js-utils "^1.0.3"

hyphenate-style-name@^1.0.2:
  version "1.0.2"
  resolved "https://registry.yarnpkg.com/hyphenate-style-name/-/hyphenate-style-name-1.0.2.tgz#31160a36930adaf1fc04c6074f7eb41465d4ec4b"

iconv-lite@~0.4.13:
  version "0.4.15"
  resolved "https://registry.yarnpkg.com/iconv-lite/-/iconv-lite-0.4.15.tgz#fe265a218ac6a57cfe854927e9d04c19825eddeb"

inline-style-prefixer@^3.0.0:
  version "3.0.2"
  resolved "https://registry.yarnpkg.com/inline-style-prefixer/-/inline-style-prefixer-3.0.2.tgz#989865e0c5de7a946acbea71e16e02741efe0dd7"
  dependencies:
    bowser "^1.6.0"
    css-in-js-utils "^1.0.3"

is-stream@^1.0.1:
  version "1.1.0"
  resolved "https://registry.yarnpkg.com/is-stream/-/is-stream-1.1.0.tgz#12d4a3dd4e68e0b79ceb8dbc84173ae80d91ca44"

isomorphic-fetch@^2.1.1:
  version "2.2.1"
  resolved "https://registry.yarnpkg.com/isomorphic-fetch/-/isomorphic-fetch-2.2.1.tgz#611ae1acf14f5e81f729507472819fe9733558a9"
  dependencies:
    node-fetch "^1.0.1"
    whatwg-fetch ">=0.10.0"

js-tokens@^3.0.0:
  version "3.0.1"
  resolved "https://registry.yarnpkg.com/js-tokens/-/js-tokens-3.0.1.tgz#08e9f132484a2c45a30907e9dc4d5567b7f114d7"

loose-envify@^1.0.0, loose-envify@^1.1.0:
  version "1.3.1"
  resolved "https://registry.yarnpkg.com/loose-envify/-/loose-envify-1.3.1.tgz#d1a8ad33fa9ce0e713d65fdd0ac8b748d478c848"
  dependencies:
    js-tokens "^3.0.0"

node-fetch@^1.0.1:
  version "1.6.3"
  resolved "https://registry.yarnpkg.com/node-fetch/-/node-fetch-1.6.3.tgz#dc234edd6489982d58e8f0db4f695029abcd8c04"
  dependencies:
    encoding "^0.1.11"
    is-stream "^1.0.1"

object-assign@^4.1.0:
  version "4.1.1"
  resolved "https://registry.yarnpkg.com/object-assign/-/object-assign-4.1.1.tgz#2109adc7965887cfc05cbbd442cac8bfbb360863"

promise@^7.1.1:
  version "7.1.1"
  resolved "https://registry.yarnpkg.com/promise/-/promise-7.1.1.tgz#489654c692616b8aa55b0724fa809bb7db49c5bf"
  dependencies:
    asap "~2.0.3"

react-fela@^4.3.2:
  version "4.3.2"
  resolved "https://registry.yarnpkg.com/react-fela/-/react-fela-4.3.2.tgz#12703afceacca43fb8dcb093917ba6793a5d149e"

react@^15.4.2:
  version "15.4.2"
  resolved "https://registry.yarnpkg.com/react/-/react-15.4.2.tgz#41f7991b26185392ba9bae96c8889e7e018397ef"
  dependencies:
    fbjs "^0.8.4"
    loose-envify "^1.1.0"
    object-assign "^4.1.0"

setimmediate@^1.0.5:
  version "1.0.5"
  resolved "https://registry.yarnpkg.com/setimmediate/-/setimmediate-1.0.5.tgz#290cbb232e306942d7d7ea9b83732ab7856f8285"

ua-parser-js@^0.7.9:
  version "0.7.12"
  resolved "https://registry.yarnpkg.com/ua-parser-js/-/ua-parser-js-0.7.12.tgz#04c81a99bdd5dc52263ea29d24c6bf8d4818a4bb"

whatwg-fetch@>=0.10.0:
  version "2.0.3"
  resolved "https://registry.yarnpkg.com/whatwg-fetch/-/whatwg-fetch-2.0.3.tgz#9c84ec2dcf68187ff00bc64e1274b442176e1c84"
