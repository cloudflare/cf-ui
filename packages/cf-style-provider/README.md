# cf-style-provider

> Cloudflare Style Provider

In order to support cf-ui (fela based components) in your project, you need to wrap your app by our StyleProvider. Check out the main README.md to see it in use. This package also provides testing utilities. Our fela setup consists of many plugins, enhancers and themes. **To ensure that everything has the same context and works properly, you should never use packages as `fela`, `fela-dom`, `react-fela` or `cf-style-const` directly**. Always use `cf-style-provider`.

## Installation

```sh
$ npm install cf-style-provider
```

## StyleProvider

```jsx
import { StyleProvider } from 'cf-style-provider';

render(
  <StyleProvider
    dev={false}
    renderer={undefined}
  >
    <YourApp />
  </StyleProvider>
);
```

### props and default values

- **dev**: *false*, in dev mode it applies more Fela plugins (validator, beautifier, monolithic)
- **renderer**: *undefined*, StyleProvider creates fela renderer internally but sometimes you might want to inject your own instance if you need to expose other renderer's APIs (e.g. for testing) or to ensure that there is only single instance of fela renderer in your application

Dev mode produces developer friendly output (CSS is formatted), validates all rules and merges atomic classNames into one. Fela automatically creates `<style>` nodes in `<head>` node and injects the styles.

## createRenderer

Or you can just get Fela's renderer (useful for server-side rendering)

```jsx
import ReactDOMServer from 'react-dom/server';
import { Provider } from 'react-fela';
import { createRenderer } from 'cf-style-provider';

// default values
const renderer = createRenderer({
  dev: false, // in dev mode it applies more Fela plugins (validator, ...)
});

const body = ReactDOMServer.renderToString(
  <Provider renderer={renderer}>
    <Main />
  </Provider>
);

renderer.renderToString(); // returns CSS as a string
```

## felaTestContext

**For testing**. HOC that passes fela renderer and base theme into the
context. Handy for testing with enzyme.

**NOTE:** felaTestContext is not exported from the index, you must provide a
deep path to acccess it.

```js
import { mount } from 'enzyme';
import { felaTestContext } from 'cf-style-provider';

test('should call onPageChange when clicking another page', () => {
  const onPageChange = createStub();

  const wrapper = mount(
    felaTestContext(
      <PaginationBuilder
        onPageChange={onPageChange}
        totalCount={20}
        page={1}
        perPage={1}
      />
    )
  );

  const items = wrapper.find('li');
  items.at(2).find('a').simulate('click');
  expect(onPageChange.called).toBeTruthy();
});
```

## felaSnapshot

**For testing**. Uses internally felaTestContext. It shallow renders the component and returns snapshots of DOM and styles as a object with properties `component` and `styles`. Use it like this:

**NOTE:** feleSnapshot is not exported from the index, you must provide a deep
path to acccess it.

```js
import React from 'react';
import { Form } from '../src/index';
import { felaSnapshot } from 'cf-style-provider';

test('should render horizontal layout', () => {
  const snapshot = felaSnapshot(
    <Form layout="horizontal" onSubmit={() => console.log('submit')}>
      Form
    </Form>
  );
  expect(snapshot.component).toMatchSnapshot();
  expect(snapshot.styles).toMatchSnapshot();
});
``` 

## Used Fela plugins

We use multiple fela plugins that extend typical CSS-in-JS syntax. Check them out for more details:

- [prefixer](https://github.com/rofrischmann/fela/tree/master/packages/fela-plugin-prefixer)
- [fallback-value](https://github.com/rofrischmann/fela/tree/master/packages/fela-plugin-fallback-value)
- [unit](https://github.com/rofrischmann/fela/tree/master/packages/fela-plugin-unit)
- [embedded](https://github.com/rofrischmann/fela/tree/master/packages/fela-plugin-embedded)
- [named-media-query](https://github.com/rofrischmann/fela/tree/master/packages/fela-plugin-named-media-query)
  - mobile
  - mobileWide
  - tablet
  - desktop
  - desktopLarge
