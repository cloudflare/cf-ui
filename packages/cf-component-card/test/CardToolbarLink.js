import React from 'react';
import { felaSnapshot } from 'cf-style-provider';
import CardToolbarLink from '../src/CardToolbarLink';

test('should render', () => {
  const snapshot = felaSnapshot(
    <CardToolbarLink id="link" isActive={false} expandable onClick={() => {}}>
      CardToolbarLink
    </CardToolbarLink>
  );
  expect(snapshot.component).toMatchSnapshot();
  expect(snapshot.styles).toMatchSnapshot();
});

test('should render with isActive', () => {
  const snapshot = felaSnapshot(
    <CardToolbarLink id="link" isActive={true} expandable onClick={() => {}}>
      CardToolbarLink
    </CardToolbarLink>
  );
  expect(snapshot.component).toMatchSnapshot();
  expect(snapshot.styles).toMatchSnapshot();
});

test('should render nonexpandable link', () => {
  const snapshot = felaSnapshot(
    <CardToolbarLink
      id="link"
      isActive={false}
      expandable={false}
      onClick={() => {}}
    >
      CardToolbarLink
    </CardToolbarLink>
  );
  expect(snapshot.component).toMatchSnapshot();
  expect(snapshot.styles).toMatchSnapshot();
});
