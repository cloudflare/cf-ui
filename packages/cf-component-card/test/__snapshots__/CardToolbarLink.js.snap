// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<a
  className="cf-link a b c d e f g h i"
  href="#!"
  id="link"
  onClick={[Function]}
  role="tab"
>
  CardToolbarLink
</a>
`;

exports[`should render 2`] = `
"
.a {
  display: inline-block
}

.b {
  margin-right: 1em
}

.c::after {
  display: inline-block
}

.d::after {
  content: '\\\\f003'
}

.e::after {
  font-family: \\"cloudflare-font\\"
}

.f::after {
  font-style: normal
}

.g::after {
  font-weight: normal
}

.h::after {
  transition: -webkit-transform 0.2s ease,transform 0.2s ease;
  -webkit-transition: -webkit-transform 0.2s ease,transform 0.2s ease;
  -moz-transition: transform 0.2s ease
}

.i::after {
  transform: initial;
  -webkit-transform: initial
}
"
`;

exports[`should render with isActive 1`] = `
<a
  className="cf-link a b c d e f g h i"
  href="#!"
  id="link"
  onClick={[Function]}
  role="tab"
>
  CardToolbarLink
</a>
`;

exports[`should render with isActive 2`] = `
"
.a {
  display: inline-block
}

.b {
  margin-right: 1em
}

.c::after {
  display: inline-block
}

.d::after {
  content: '\\\\f003'
}

.e::after {
  font-family: \\"cloudflare-font\\"
}

.f::after {
  font-style: normal
}

.g::after {
  font-weight: normal
}

.h::after {
  transition: -webkit-transform 0.2s ease,transform 0.2s ease;
  -webkit-transition: -webkit-transform 0.2s ease,transform 0.2s ease;
  -moz-transition: transform 0.2s ease
}

.i::after {
  transform: rotate(90deg);
  -webkit-transform: rotate(90deg)
}
"
`;
